firestore/firestore_snippets/listen_test.go

```diff
@@ -101,29 +101,31 @@ func TestListenChanges(t *testing.T) {
 	ctx, cancel := context.WithTimeout(ctx, duration)
 	defer cancel()
 
-	buf := &bytes.Buffer{}
-	c := make(chan *bytes.Buffer)
-	go func() {
-		defer close(c)
-		err := listenChanges(ctx, buf, projectID, collection)
-		if err != nil {
-			t.Errorf("listenChanges: %v", err)
+	testutil.Retry(t, 10, 2*time.Second, func(r *testutil.R) {
+		buf := &bytes.Buffer{}
+		c := make(chan *bytes.Buffer)
+		go func() {
+			defer close(c)
+			err := listenChanges(ctx, buf, projectID, collection)
+			if err != nil {
+				r.Errorf("listenChanges: %v", err)
+			}
+			c <- buf
+		}()
+		// Add some changes to data in parallel.
+		time.Sleep(time.Second)
+		var pop int64 = 3900000
+		if _, err := client.Collection(collection).Doc("LA").Update(ctx, []firestore.Update{
```

Comment:
Is there anything stateful about this? That is, if you run the test twice, will the change look the same on the listener end both times? Just wanted to confirm.

Reply:
Listeners report on changes since listener start. so starting the func to listen, then updating the document, should result in an update containing population. This should happen pretty rapidly.

It is possible that we also could add a delay before line 125, ensuring that the listener had a chance to observe the data update.

in essence, I think I may be able to scope to just ln123-128

Reply:
After further thought I am fairly sure the issue is `<-c` is unblocking due to reading the initial state, not also including changes (update events). We can retry the reading of the buffer a couple of times to ensure we are in a steady state.